pkg/chartsync/git.go

```diff
@@ -0,0 +1,337 @@
+package chartsync
+
+import (
+	"context"
+	"errors"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/weaveworks/flux/git"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/fluxcd/helm-operator/pkg/apis/helm.fluxcd.io/v1"
+	lister "github.com/fluxcd/helm-operator/pkg/client/listers/helm.fluxcd.io/v1"
+)
+
+// Various (final) errors.
+var (
+	ErrReleasesForMirror = errors.New("failed to get HelmRelease resources for mirror")
+	ErrNoMirror          = errors.New("no existing git mirror found")
+	ErrMirrorSync        = errors.New("failed syncing git mirror")
+)
+
+// ReleaseQueue is an add-only `workqueue.RateLimitingInterface`
+type ReleaseQueue interface {
+	AddRateLimited(item interface{})
+}
+
+// GitConfig holds the configuration for git operations.
+type GitConfig struct {
+	GitTimeout      time.Duration
+	GitPollInterval time.Duration
+}
+
+// GitChartDir holds a `git.Export` of a chart repository,
+// and the commit hash of the HEAD of this export.
+type GitChartDir struct {
```

Comment:
I was going to ask for a name change (something like `GitChartWorkingDir`) but, after seeing its use I think its overkill to have an explicit structure. I would recommend getting rid of it and returning an `Export` and the `Head` explicitly (`ChartPath` can be implemented by the consumer).